File: 0091 Decode Ways.py
class Solution:
    def numDecodings(self, s: str) -> int:
        if not s: return 0
        l = len(s)
        dp = [0] * (l+1)
        dp[0], dp[1] = 1, 1 if s[0] != "0" else 0
        prev, curr = "0", s[0]
        for i in range(2, l + 1):
            prev = curr
            curr = s[i-1]
            if curr != "0":
                dp[i] += dp[i-1]
            if prev == "1" or prev == "2" and curr <= "6":
                dp[i] += dp[i-2]
        return dp[l]

File: 0091 Decode Ways.java
class Solution {
    public int numDecodings(String s) {
        if (s.length() == 0) return 0;
        int l = s.length();
        int dp[] = new int[l+1];
        dp[0] = 1; // Empty string
        dp[1] = s.charAt(0) != '0' ? 1 : 0;
        char prev = '0', curr = s.charAt(0);
        for (int i = 2; i <= l; ++i) {
            prev = curr;
            curr = s.charAt(i-1);
            if (curr != '0')
                dp[i] += dp[i-1];
            if (prev == '1' || (prev == '2' && curr <= '6'))
                dp[i] += dp[i-2];
        }
        return dp[l];
    }
}